File: zz/BottomBar.swift
import SwiftUI
import WebKit
#if canImport(UIKit)
import UIKit
#elseif canImport(AppKit)
import AppKit
#endif

struct BottomBar: View {
    @Binding var draft: String
    @FocusState.Binding var urlFocused: Bool
    @Binding var selectedSuggestionIndex: Int?
    let matches: [OmniboxItem]
    @Binding var sidebarPresented: Bool
    @Binding var settingsPresented: Bool
    @Binding var historyPresented: Bool
    let onSaveLayout: () -> Void
    let onApplyLayout: (LayoutPreset) -> Void
    let onDeleteLayout: (UUID) -> Void
    let onOutsideURLBarInteraction: () -> Void
    let onCommit: (String) -> Void
    let onSelect: (OmniboxItem) -> Void

    @Environment(BrowserStore.self) private var store
    @Environment(\.horizontalSizeClass) private var horizontalSizeClass

    var body: some View {
        let tab = store.focusedTab
        HStack(spacing: 6) {
            navButtons(tab: tab)
            URLBar(
                text: $draft,
                focused: $urlFocused,
                findEnabled: tab != nil && !(tab?.isBlank ?? true),
                onFind: { store.findInFocused() },
                onSubmit: submit
            )
                .onKeyPress(.downArrow) { moveSelection(+1) }
                .onKeyPress(.upArrow)   { moveSelection(-1) }
                .onKeyPress(.escape) {
                    if selectedSuggestionIndex != nil {
                        selectedSuggestionIndex = nil
                        return .handled
                    }
                    if urlFocused {
                        urlFocused = false
                        return .handled
                    }
                    return .ignored
                }
                .onChange(of: draft) { _, _ in
                    selectedSuggestionIndex = nil
                }
                .layoutPriority(1)
            if isCompact {
                MoreMenu(
                    tab: tab,
                    isCompact: isCompact,
                    settingsPresented: $settingsPresented,
                    historyPresented: $historyPresented,
                    onSaveLayout: onSaveLayout,
                    onApplyLayout: onApplyLayout,
                    onDeleteLayout: onDeleteLayout,
                    onOutsideURLBarInteraction: onOutsideURLBarInteraction
                )
                BarIconButton(name: "sidebar.right",
                              enabled: !store.parked.isEmpty,
                              action: { runOutsideURLBarInteraction { sidebarPresented = true } },
                              help: "Sidebar")
            } else {
                actionButtons(tab: tab)
            }
        }
        .padding(.horizontal, 10)
        .padding(.top, 6)
        .padding(.bottom, 2)
        .background(.bar, ignoresSafeAreaEdges: .bottom)
        .overlay(alignment: .top) {
            Rectangle().fill(.separator.opacity(0.28)).frame(height: 0.5)
        }
    }

    @ViewBuilder
    private func navButtons(tab: Tab?) -> some View {
        HStack(spacing: 2) {
            HistoryMenu(
                tab: tab,
                items: (tab?.backList ?? []).reversed(),
                icon: "chevron.backward",
                help: "Back (⌘[)",
                primaryAction: { tab?.goBack() },
                enabled: tab?.canGoBack ?? false,
                onInteraction: onOutsideURLBarInteraction
            )
            HistoryMenu(
                tab: tab,
                items: Array(tab?.forwardList ?? []),
                icon: "chevron.forward",
                help: "Forward (⌘])",
                primaryAction: { tab?.goForward() },
                enabled: tab?.canGoForward ?? false,
                onInteraction: onOutsideURLBarInteraction
            )
            ReloadControl(tab: tab, onInteraction: onOutsideURLBarInteraction)
        }
    }

    @ViewBuilder
    private func actionButtons(tab: Tab?) -> some View {
        HStack(spacing: 2) {
            BarIconButton(name: store.zoomedTabID == nil
                            ? "arrow.up.left.and.arrow.down.right"
                            : "arrow.down.right.and.arrow.up.left",
                          enabled: tab != nil,
                          action: { runOutsideURLBarInteraction { store.toggleZoom() } },
                          help: store.zoomedTabID == nil
                            ? "Zoom focused tile (⌃⌥⌘F)"
                            : "Restore layout (⌃⌥⌘F)")
            BarIconButton(name: "tray.and.arrow.down",
                          enabled: !(tab?.isBlank ?? true),
                          action: {
                              runOutsideURLBarInteraction {
                                  if let id = store.focusedTabID { store.park(id) }
                              }
                          },
                          help: "Park (⌥⌘P)")
            BarIconButton(name: "rectangle.split.1x2", enabled: store.canSplitSelection,
                          action: {
                              runOutsideURLBarInteraction {
                                  store.splitSelection(axis: .horizontal)
                              }
                          },
                          help: "Horizontal Split (⌘\\)")
            BarIconButton(name: "rectangle.split.2x1", enabled: store.canSplitSelection,
                          action: {
                              runOutsideURLBarInteraction {
                                  store.splitSelection(axis: .vertical)
                              }
                          },
                          help: "Vertical Split (⇧⌘\\)")
            MoreMenu(
                tab: tab,
                isCompact: false,
                settingsPresented: $settingsPresented,
                historyPresented: $historyPresented,
                onSaveLayout: onSaveLayout,
                onApplyLayout: onApplyLayout,
                onDeleteLayout: onDeleteLayout,
                onOutsideURLBarInteraction: onOutsideURLBarInteraction
            )
            BarIconButton(name: "xmark", enabled: tab != nil,
                          action: {
                              runOutsideURLBarInteraction {
                                  if let id = store.focusedTabID { store.close(id) }
                              }
                          },
                          help: "Close tile (⌘W)")
        }
    }

    private func submit() {
        // Keyboard submit must route on kind exactly like click-select so an
        // open-tab row focuses rather than reloads.
        if let idx = selectedSuggestionIndex, idx >= 0, idx < matches.count {
            onSelect(matches[idx])
        } else {
            onCommit(draft)
        }
    }

    private func moveSelection(_ delta: Int) -> KeyPress.Result {
        guard !matches.isEmpty else { return .ignored }
        let current = selectedSuggestionIndex ?? -1
        let next = current + delta
        let clamped = max(0, min(matches.count - 1, next))
        selectedSuggestionIndex = clamped
        return .handled
    }

    private var isCompact: Bool {
        #if os(macOS)
        return false
        #else
        return horizontalSizeClass == .compact
        #endif
    }

    private func runOutsideURLBarInteraction(_ action: () -> Void) {
        onOutsideURLBarInteraction()
        action()
    }
}

private struct MoreMenu: View {
    let tab: Tab?
    let isCompact: Bool
    @Binding var settingsPresented: Bool
    @Binding var historyPresented: Bool
    let onSaveLayout: () -> Void
    let onApplyLayout: (LayoutPreset) -> Void
    let onDeleteLayout: (UUID) -> Void
    let onOutsideURLBarInteraction: () -> Void

    @Environment(BrowserStore.self) private var store
    @Environment(HistoryStore.self) private var history
    @Environment(LayoutPresetStore.self) private var layouts
    @Environment(\.openWindow) private var openWindow
    @AppStorage(BrowserPreferences.recordHistoryKey) private var recordHistory = true

    var body: some View {
        Menu {
            Button {
                runOutsideURLBarInteraction {
                    openWindow(value: WindowID())
                }
            } label: {
                Label("New Window", systemImage: "macwindow.badge.plus")
            }

            if tab != nil {
                Divider()
                focusedPaneActions
                Divider()
            }

            layoutMenu
            layoutsMenu
            privacyHistoryMenu

            Divider()

            Button {
                runOutsideURLBarInteraction {
                    historyPresented = true
                }
            } label: {
                Label("History", systemImage: "clock.arrow.circlepath")
            }

            Button {
                runOutsideURLBarInteraction {
                    settingsPresented = true
                }
            } label: {
                Label("Settings", systemImage: "gearshape")
            }
        } label: {
            Image(systemName: "ellipsis")
                .font(.system(size: 14, weight: .medium))
                .frame(width: 30, height: 30)
                .contentShape(.rect)
        }
        .menuStyle(.button)
        .buttonStyle(.plain)
        .menuIndicator(.hidden)
        .menuOrder(.fixed)
        .help("More")
    }

    @ViewBuilder
    private var focusedPaneActions: some View {
        if let tab, let tabID = store.focusedTabID {
            let actions = TileMenuActions(isBlank: tab.isBlank)

            if actions.canDuplicate {
                Button {
                    runOutsideURLBarInteraction {
                        store.split(tabID, axis: .vertical, side: .after,
                                    loadURL: tab.currentURL)
                    }
                } label: {
                    Label("Duplicate Tile", systemImage: "plus.rectangle.on.rectangle")
                }
            }

            if actions.canCopyURL {
                Button {
                    runOutsideURLBarInteraction {
                        copyURL(tab.currentURL)
                    }
                } label: {
                    Label("Copy URL", systemImage: "doc.on.doc")
                }
            }

            if actions.canReload {
                Button {
                    runOutsideURLBarInteraction {
                        tab.reload()
                    }
                } label: {
                    Label("Reload", systemImage: "arrow.clockwise")
                }
            }

            if actions.canRequestDesktopSite {
                Toggle(isOn: Binding(
                    get: { tab.requestsDesktopSite },
                    set: {
                        onOutsideURLBarInteraction()
                        tab.requestsDesktopSite = $0
                    }
                )) {
                    Label("Request Desktop Site", systemImage: "desktopcomputer")
                }
            }

            if actions.canSuspendMedia {
                Toggle(isOn: Binding(
                    get: { tab.isMediaSuspended },
                    set: {
                        onOutsideURLBarInteraction()
                        tab.isMediaSuspended = $0
                    }
                )) {
                    Label(tab.isMediaSuspended ? "Resume Media" : "Suspend Media",
                          systemImage: tab.isMediaSuspended ? "play.circle" : "pause.circle")
                }
            }

            if actions.canPark {
                Button {
                    runOutsideURLBarInteraction {
                        store.park(tabID)
                    }
                } label: {
                    Label("Park", systemImage: "tray.and.arrow.down")
                }
            }

            if actions.canPark || actions.canCopyURL {
                Divider()
            }

            Button(role: .destructive) {
                runOutsideURLBarInteraction {
                    store.close(tabID)
                }
            } label: {
                Label("Close", systemImage: "xmark")
            }
        }
    }

    @ViewBuilder
    private var layoutMenu: some View {
        Menu {
            Button {
                runOutsideURLBarInteraction {
                    store.selectParentGroup()
                }
            } label: {
                Label("Select Parent Group", systemImage: "square.stack.3d.up")
            }
            .disabled(!store.canSelectParentGroup)

            Button {
                runOutsideURLBarInteraction {
                    store.equalizeSelectedGroup()
                }
            } label: {
                Label("Equalize Group", systemImage: "rectangle.split.2x2")
            }
            .disabled(!store.canTransformSelectedGroup)

            Button {
                runOutsideURLBarInteraction {
                    store.rotateSelectedGroup()
                }
            } label: {
                Label("Rotate Group", systemImage: "rotate.right")
            }
            .disabled(!store.canTransformSelectedGroup)

            if isCompact {
                Divider()

                Button {
                    runOutsideURLBarInteraction {
                        store.splitSelection(axis: .horizontal)
                    }
                } label: {
                    Label("Horizontal Split", systemImage: "rectangle.split.1x2")
                }
                .disabled(!store.canSplitSelection)

                Button {
                    runOutsideURLBarInteraction {
                        store.splitSelection(axis: .vertical)
                    }
                } label: {
                    Label("Vertical Split", systemImage: "rectangle.split.2x1")
                }
                .disabled(!store.canSplitSelection)
            }
        } label: {
            Label("Layout", systemImage: "rectangle.3.group")
        }
    }

    @ViewBuilder
    private var layoutsMenu: some View {
        Menu {
            Button {
                runOutsideURLBarInteraction {
                    onSaveLayout()
                }
            } label: {
                Label("Save Current Layout…", systemImage: "square.and.arrow.down")
            }

            if !layouts.presets.isEmpty {
                Divider()

                ForEach(layouts.presets) { preset in
                    Menu {
                        Button {
                            runOutsideURLBarInteraction {
                                onApplyLayout(preset)
                            }
                        } label: {
                            Label("Apply", systemImage: "rectangle.on.rectangle")
                        }
                        Button(role: .destructive) {
                            runOutsideURLBarInteraction {
                                onDeleteLayout(preset.id)
                            }
                        } label: {
                            Label("Delete", systemImage: "trash")
                        }
                    } label: {
                        Text(preset.name)
                    }
                }
            }
        } label: {
            Label("Layouts", systemImage: "square.grid.2x2")
        }
    }

    @ViewBuilder
    private var privacyHistoryMenu: some View {
        Menu {
            Toggle("Record History", isOn: Binding(
                get: { recordHistory },
                set: {
                    onOutsideURLBarInteraction()
                    recordHistory = $0
                }
            ))

            Button(role: .destructive) {
                runOutsideURLBarInteraction {
                    history.clear()
                }
            } label: {
                Label("Clear History", systemImage: "trash")
            }
        } label: {
            Label("Privacy & History", systemImage: "hand.raised")
        }
    }

    private func copyURL(_ urlString: String) {
        #if canImport(UIKit)
        UIPasteboard.general.string = urlString
        #elseif canImport(AppKit)
        let pasteboard = NSPasteboard.general
        pasteboard.clearContents()
        pasteboard.setString(urlString, forType: .string)
        #endif
    }

    private func runOutsideURLBarInteraction(_ action: () -> Void) {
        onOutsideURLBarInteraction()
        action()
    }
}

struct TileMenuActions: Equatable {
    let canDuplicate: Bool
    let canCopyURL: Bool
    let canReload: Bool
    let canPark: Bool
    let canRequestDesktopSite: Bool
    let canSuspendMedia: Bool

    init(isBlank: Bool) {
        let hasContent = !isBlank
        canDuplicate = hasContent
        canCopyURL = hasContent
        canReload = hasContent
        canPark = hasContent
        canRequestDesktopSite = hasContent
        canSuspendMedia = hasContent
    }
}

private struct ReloadControl: View {
    let tab: Tab?
    let onInteraction: () -> Void
    @State private var suppressNextTap = false
    @State private var suppressResetTask: Task<Void, Never>?

    var body: some View {
        if tab?.isLoading == true {
            BarIconButton(
                name: "xmark",
                enabled: tab != nil,
                action: {
                    onInteraction()
                    tab?.stop()
                },
                help: "Stop"
            )
        } else {
            Button {
                onInteraction()
                if suppressNextTap {
                    suppressNextTap = false
                } else {
                    tab?.reload()
                }
            } label: {
                iconLabel
            }
            .buttonStyle(.plain)
            .disabled(tab == nil)
            .opacity(tab == nil ? 0.35 : 1)
            .simultaneousGesture(
                LongPressGesture(minimumDuration: 0.45)
                    .onEnded { _ in
                        guard tab != nil else { return }
                        onInteraction()
                        suppressNextTap = true
                        tab?.forceReload()
                        suppressResetTask?.cancel()
                        suppressResetTask = Task { @MainActor in
                            try? await Task.sleep(for: .milliseconds(700))
                            guard !Task.isCancelled else { return }
                            suppressNextTap = false
                        }
                    }
            )
            .help("Reload (⌘R), Force Reload (long-press or ⇧⌘R)")
            .onChange(of: tab?.isLoading) { _, _ in
                suppressResetTask?.cancel()
                suppressResetTask = nil
                suppressNextTap = false
            }
            .onChange(of: tab?.id) { _, _ in
                suppressResetTask?.cancel()
                suppressResetTask = nil
                suppressNextTap = false
            }
            .onDisappear {
                suppressResetTask?.cancel()
                suppressResetTask = nil
            }
        }
    }

    private var iconLabel: some View {
        Image(systemName: "arrow.clockwise")
            .font(.system(size: 14, weight: .medium))
            .frame(width: 30, height: 30)
            .contentShape(.rect)
    }
}

private struct HistoryMenu: View {
    let tab: Tab?
    let items: [WKBackForwardListItem]
    let icon: String
    let help: String
    let primaryAction: () -> Void
    let enabled: Bool
    let onInteraction: () -> Void

    var body: some View {
        Group {
            if enabled && !items.isEmpty {
                Menu {
                    ForEach(items.prefix(25), id: \.self) { item in
                        Button {
                            onInteraction()
                            tab?.go(to: item)
                        } label: {
                            Text(label(for: item))
                        }
                    }
                } label: {
                    iconLabel
                } primaryAction: {
                    onInteraction()
                    primaryAction()
                }
                .menuStyle(.button)
                .buttonStyle(.plain)
                .menuIndicator(.hidden)
                .menuOrder(.fixed)
            } else {
                Button {
                    onInteraction()
                    primaryAction()
                } label: {
                    iconLabel
                }
                    .buttonStyle(.plain)
                    .disabled(!enabled)
            }
        }
        .opacity(enabled ? 1 : 0.35)
        .help(help)
    }

    private var iconLabel: some View {
        Image(systemName: icon)
            .font(.system(size: 14, weight: .medium))
            .frame(width: 30, height: 30)
            .contentShape(.rect)
    }

    private func label(for item: WKBackForwardListItem) -> String {
        if let title = item.title, !title.isEmpty { return title }
        return item.url.host(percentEncoded: false) ?? item.url.absoluteString
    }
}

private struct BarIconButton: View {
    let name: String
    let enabled: Bool
    let action: () -> Void
    var help: String = ""

    var body: some View {
        Button(action: action) {
            Image(systemName: name)
                .font(.system(size: 14, weight: .medium))
                .frame(width: 30, height: 30)
                .contentShape(.rect)
        }
        .buttonStyle(.plain)
        .disabled(!enabled)
        .opacity(enabled ? 1 : 0.35)
        .help(help)
    }
}

File: android/app/src/main/java/surf/zz/ui/bottombar/BottomBar.kt
package surf.zz.ui.bottombar

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxWithConstraints
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material.icons.automirrored.filled.ArrowForward
import androidx.compose.material.icons.automirrored.filled.ViewSidebar
import androidx.compose.material.icons.filled.Close
import androidx.compose.material.icons.filled.HorizontalSplit
import androidx.compose.material.icons.filled.Inbox
import androidx.compose.material.icons.filled.VerticalSplit
import androidx.compose.material.icons.filled.ZoomInMap
import androidx.compose.material.icons.filled.ZoomOutMap
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.input.key.Key
import androidx.compose.ui.input.key.KeyEventType
import androidx.compose.ui.input.key.key
import androidx.compose.ui.input.key.onPreviewKeyEvent
import androidx.compose.ui.input.key.type
import androidx.compose.ui.platform.LocalFocusManager
import androidx.compose.ui.unit.dp
import surf.zz.browser.tab.Tab
import surf.zz.layout.BspNode
import surf.zz.omnibox.OmniboxItem
import surf.zz.store.LayoutPreset
import surf.zz.ui.LocalBrowserStore
import surf.zz.ui.omnibox.UrlBar
import java.util.UUID

/**
 * The bottom toolbar: navigation controls + the omnibox [UrlBar] + action buttons,
 * collapsing to a [MoreMenu] + sidebar button on a compact width.
 *
 * Faithful port of `BottomBar.swift`'s `struct BottomBar`. State and callbacks are
 * hoisted (the iOS `@Binding` / closures), the [surf.zz.store.BrowserStore] comes from
 * a CompositionLocal (the iOS `@Environment`), and the keyboard omnibox handling
 * (down/up/escape/enter) is reattached via [Modifier.onPreviewKeyEvent] on the URL
 * field. The nav cluster ([HistoryMenu] / [ReloadControl]) and the overflow
 * [MoreMenu] are the shared bottombar composables.
 *
 * DEVIATIONS from iOS (documented):
 *  - `horizontalSizeClass == .compact` becomes a width threshold measured by
 *    [BoxWithConstraints]: a bar narrower than [COMPACT_WIDTH] uses the compact
 *    layout. macOS forced `isCompact == false`; on a phone/tablet the width decides.
 *  - `.background(.bar, ignoresSafeAreaEdges: .bottom)` becomes a [Surface] tinted
 *    [MaterialTheme.colorScheme.surface]; edge-to-edge bottom-inset handling is the
 *    host screen's responsibility (ANDROID_ARCH.md §13).
 *  - SF Symbols map to the nearest Material icons (ANDROID_ARCH.md §11); the split
 *    glyphs use [Icons.Filled.HorizontalSplit] / [Icons.Filled.VerticalSplit].
 *  - `.help(...)` tooltips become content descriptions; the keyboard-shortcut hints
 *    in the help strings are dropped on a touch UI.
 */
@Composable
fun BottomBar(
    draft: String,
    onDraftChange: (String) -> Unit,
    urlFocused: Boolean,
    onFocusChange: (Boolean) -> Unit,
    selectedSuggestionIndex: Int?,
    setSelectedSuggestionIndex: (Int?) -> Unit,
    matches: List<OmniboxItem>,
    sidebarPresented: Boolean,
    onSidebarPresentedChange: (Boolean) -> Unit,
    settingsPresented: Boolean,
    onSettingsPresentedChange: (Boolean) -> Unit,
    historyPresented: Boolean,
    onHistoryPresentedChange: (Boolean) -> Unit,
    onSaveLayout: () -> Unit,
    onApplyLayout: (LayoutPreset) -> Unit,
    onDeleteLayout: (UUID) -> Unit,
    onOutsideUrlBarInteraction: () -> Unit,
    onCommit: (String) -> Unit,
    onSelect: (OmniboxItem) -> Unit,
    onFind: () -> Unit,
) {
    val store = LocalBrowserStore.current
    val focusManager = LocalFocusManager.current
    val tab = store.focusedTab

    // submit(): keyboard submit must route on kind exactly like click-select so an
    // open-tab row focuses rather than reloads (iOS BottomBar.submit()).
    fun submit() {
        val idx = selectedSuggestionIndex
        if (idx != null && idx >= 0 && idx < matches.size) {
            onSelect(matches[idx])
        } else {
            onCommit(draft)
        }
    }

    // moveSelection(delta): clamp into [0, matches.size-1], seeding from -1 when no
    // row is selected. Returns true when consumed (iOS KeyPress.Result).
    fun moveSelection(delta: Int): Boolean {
        if (matches.isEmpty()) return false
        val current = selectedSuggestionIndex ?: -1
        val next = current + delta
        val clamped = next.coerceIn(0, matches.size - 1)
        setSelectedSuggestionIndex(clamped)
        return true
    }

    Surface(
        color = MaterialTheme.colorScheme.surface,
        modifier = Modifier.fillMaxWidth(),
    ) {
        Box {
            BoxWithConstraints {
                val isCompact = maxWidth < COMPACT_WIDTH

                Row(
                    modifier = Modifier
                        .fillMaxWidth()
                        // contentShape(.rect) + onTapGesture { } : swallow background taps
                        // so they do not fall through to the panes behind the bar.
                        .clickable(
                            interactionSource = remember { MutableInteractionSource() },
                            indication = null,
                            onClick = {},
                        )
                        .padding(horizontal = 10.dp)
                        .padding(top = 6.dp, bottom = 2.dp),
                    horizontalArrangement = Arrangement.spacedBy(6.dp),
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    NavControls(
                        tab = tab,
                        onOutsideUrlBarInteraction = onOutsideUrlBarInteraction,
                    )

                    UrlBar(
                        text = draft,
                        onTextChange = { newDraft ->
                            onDraftChange(newDraft)
                            // onChange(of: draft): clear selection on edit.
                            setSelectedSuggestionIndex(null)
                        },
                        focused = urlFocused,
                        onFocusChange = onFocusChange,
                        findEnabled = tab != null && !tab.isBlank,
                        onFind = onFind,
                        onSubmit = { submit() },
                        modifier = Modifier
                            .weight(1f)
                            .onPreviewKeyEvent { event ->
                                if (event.type != KeyEventType.KeyDown) return@onPreviewKeyEvent false
                                when (event.key) {
                                    Key.DirectionDown -> moveSelection(+1)
                                    Key.DirectionUp -> moveSelection(-1)
                                    Key.Enter, Key.NumPadEnter -> {
                                        submit()
                                        true
                                    }
                                    Key.Escape -> {
                                        // escape: clear selection first, then drop focus.
                                        if (selectedSuggestionIndex != null) {
                                            setSelectedSuggestionIndex(null)
                                            true
                                        } else if (urlFocused) {
                                            focusManager.clearFocus()
                                            onFocusChange(false)
                                            true
                                        } else {
                                            false
                                        }
                                    }
                                    else -> false
                                }
                            },
                    )

                    if (isCompact) {
                        MoreMenu(
                            tab = tab,
                            isCompact = true,
                            onSettingsPresented = { onSettingsPresentedChange(true) },
                            onHistoryPresented = { onHistoryPresentedChange(true) },
                            onSaveLayout = onSaveLayout,
                            onApplyLayout = onApplyLayout,
                            onDeleteLayout = onDeleteLayout,
                            onOutsideUrlBarInteraction = onOutsideUrlBarInteraction,
                        )
                        BarIconButton(
                            icon = Icons.AutoMirrored.Filled.ViewSidebar,
                            enabled = store.parked.isNotEmpty(),
                            onClick = {
                                runOutsideUrlBarInteraction(onOutsideUrlBarInteraction) {
                                    onSidebarPresentedChange(true)
                                }
                            },
                            help = "Sidebar",
                        )
                    } else {
                        ActionButtons(
                            tab = tab,
                            onSettingsPresentedChange = onSettingsPresentedChange,
                            onHistoryPresentedChange = onHistoryPresentedChange,
                            onSaveLayout = onSaveLayout,
                            onApplyLayout = onApplyLayout,
                            onDeleteLayout = onDeleteLayout,
                            onOutsideUrlBarInteraction = onOutsideUrlBarInteraction,
                        )
                    }
                }
            }

            // overlay(alignment: .top) { Rectangle .separator.opacity(0.28) 0.5pt }
            Box(
                modifier = Modifier
                    .fillMaxWidth()
                    .height(0.5.dp)
                    .align(Alignment.TopCenter)
                    .background(MaterialTheme.colorScheme.onSurfaceVariant.copy(alpha = 0.28f)),
            )
        }
    }
}

/** Width below which the bottom bar collapses to the compact layout. */
private val COMPACT_WIDTH = 600.dp

/**
 * Back / forward / reload cluster. Port of `BottomBar.navButtons(tab:)`.
 *
 * Delegates to the shared [HistoryMenu] (tap = navigate one step, long-press = a
 * history dropdown) and [ReloadControl] (stop while loading, else reload /
 * long-press force-reload). The iOS `HStack(spacing: 2)` becomes a 2.dp-spaced [Row].
 */
@Composable
private fun NavControls(
    tab: Tab?,
    onOutsideUrlBarInteraction: () -> Unit,
) {
    Row(
        horizontalArrangement = Arrangement.spacedBy(2.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        HistoryMenu(
            tab = tab,
            items = tab?.backList ?: emptyList(),
            icon = Icons.AutoMirrored.Filled.ArrowBack,
            help = "Back",
            primaryAction = { tab?.goBack() },
            enabled = tab?.canGoBack ?: false,
            onInteraction = onOutsideUrlBarInteraction,
        )
        HistoryMenu(
            tab = tab,
            items = tab?.forwardList ?: emptyList(),
            icon = Icons.AutoMirrored.Filled.ArrowForward,
            help = "Forward",
            primaryAction = { tab?.goForward() },
            enabled = tab?.canGoForward ?: false,
            onInteraction = onOutsideUrlBarInteraction,
        )
        ReloadControl(tab = tab, onInteraction = onOutsideUrlBarInteraction)
    }
}

/**
 * Full action-button row for the wide layout. Port of `BottomBar.actionButtons(tab:)`:
 * zoom / park / horizontal-split / vertical-split / [MoreMenu] / close.
 */
@Composable
private fun ActionButtons(
    tab: Tab?,
    onSettingsPresentedChange: (Boolean) -> Unit,
    onHistoryPresentedChange: (Boolean) -> Unit,
    onSaveLayout: () -> Unit,
    onApplyLayout: (LayoutPreset) -> Unit,
    onDeleteLayout: (UUID) -> Unit,
    onOutsideUrlBarInteraction: () -> Unit,
) {
    val store = LocalBrowserStore.current
    Row(
        horizontalArrangement = Arrangement.spacedBy(2.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        BarIconButton(
            icon = if (store.zoomedTabID == null) Icons.Filled.ZoomOutMap else Icons.Filled.ZoomInMap,
            enabled = tab != null,
            onClick = {
                runOutsideUrlBarInteraction(onOutsideUrlBarInteraction) { store.toggleZoom() }
            },
            help = if (store.zoomedTabID == null) "Zoom focused tile" else "Restore layout",
        )
        BarIconButton(
            icon = Icons.Filled.Inbox,
            enabled = !(tab?.isBlank ?: true),
            onClick = {
                runOutsideUrlBarInteraction(onOutsideUrlBarInteraction) {
                    store.focusedTabID?.let { store.park(it) }
                }
            },
            help = "Park",
        )
        BarIconButton(
            icon = Icons.Filled.HorizontalSplit,
            enabled = store.canSplitSelection,
            onClick = {
                runOutsideUrlBarInteraction(onOutsideUrlBarInteraction) {
                    store.splitSelection(axis = BspNode.Axis.HORIZONTAL)
                }
            },
            help = "Horizontal Split",
        )
        BarIconButton(
            icon = Icons.Filled.VerticalSplit,
            enabled = store.canSplitSelection,
            onClick = {
                runOutsideUrlBarInteraction(onOutsideUrlBarInteraction) {
                    store.splitSelection(axis = BspNode.Axis.VERTICAL)
                }
            },
            help = "Vertical Split",
        )
        MoreMenu(
            tab = tab,
            isCompact = false,
            onSettingsPresented = { onSettingsPresentedChange(true) },
            onHistoryPresented = { onHistoryPresentedChange(true) },
            onSaveLayout = onSaveLayout,
            onApplyLayout = onApplyLayout,
            onDeleteLayout = onDeleteLayout,
            onOutsideUrlBarInteraction = onOutsideUrlBarInteraction,
        )
        BarIconButton(
            icon = Icons.Filled.Close,
            enabled = tab != null,
            onClick = {
                runOutsideUrlBarInteraction(onOutsideUrlBarInteraction) {
                    store.focusedTabID?.let { store.close(it) }
                }
            },
            help = "Close tile",
        )
    }
}

/**
 * Shared inline helper. Port of the file-private
 * `runOutsideURLBarInteraction(_:_:)`: notify that an interaction landed outside the
 * URL bar, then run the action. Keeps the call order identical to iOS.
 */
private inline fun runOutsideUrlBarInteraction(
    onOutsideUrlBarInteraction: () -> Unit,
    action: () -> Unit,
) {
    onOutsideUrlBarInteraction()
    action()
}
